### ISSUE TITLE
useNavigationPrompt does not work with Back button

### ISSUE THREAD

**Issue by roypearce** [kyeotic/raviger#81]
Thanks for the work & maintenance you put into this router. I have found an issue where I have a complex form, and need to prompt the user if they try to navigate away with unsaved changes. That works great for events that trigger the `beforeunload` event (hard page reload, navigate to a different site outside the SPA) or that use raviger's `navigate`, but it doesn't work with the browsers Back button. 

Has there been any consideration for supporting the interception of the back button within the SPA?

**Comment by kyeotic** [comment]
Hey @roypearce 👋 

This is absolutely something I would like to support. I didn't know the back button wasn't working, I will have to do some more tests. Have you tried this in multiple browsers? If not, can you share the browser and OS you are seeing this behavior on?

Also, if you know what it would take to add this support please let me know, I'm happy to do the work once I know what it is.

**Comment by roypearce** [comment]
Hey @kyeotic I appreciate your willingness to support this! 

I believe the back button is going to trigger a `popstate` event (it is not cancelable) which is handled here https://github.com/kyeotic/raviger/blob/74e3b87eaf108d0950ab09f368f1f91f23345dcb/src/path.js#L83 so at this point in code it would have to interface with the `handler` for `useNavigationPrompt` here https://github.com/kyeotic/raviger/blob/74e3b87eaf108d0950ab09f368f1f91f23345dcb/src/navigate.js#L33 to trigger the prompt if it needs to and hold off on triggering the SPA view from changing unless the user confirms they want to navigate.

But at that point the history stack will be modified by the browser, so if the user "cancels" the navigation through the prompt then the history entry that was removed by the browser will need to be re-added so the history will be in the correct state.

That's what I understand from my 30 minutes of research :)

**Comment by kyeotic** [comment]
Yea it looks like, since this event isn't cancellable, it must be actually _intercepted_, as in the browser's `history.pushState` function needs to be override. I've tried to avoid doing this in order to reduce conflicts with other code on the page (such as another instance of raviger), but I don't see another option. I will need some time to think about this.

**Comment by roypearce** [comment]
I don't think you need to override the `history.pushState` as much as work around it. Pardon the pseudo-code.
```
const onPopState = useCallback(() => {
  // No predicate defaults true
  if (isActive !== undefined && !isPredicateActive(isActive)) return
  if (shouldCancelNavigation()) {
    window.history.pushState(null, null, url)
    return
  }
  setRef.current(getFormattedPath(basePath))
}, [isActive, basePath])
```

**Comment by kyeotic** [comment]
Wont that cause the navigation to go forward then back? Does that actually "cancel" the navigation?

**Comment by roypearce** [comment]
[This mdn page](https://developer.mozilla.org/en-US/docs/Web/API/Window/popstate_event#when_popstate_is_sent) (also scroll up to read the blue `Note` section). In theory when the pseudo code is executed, the history change should have happened, so if it is canceled then it just puts the history back which the browser removed. If it doesn't work like that in practice then it can be wrapped in a `setTimeout` to put the `pushState` at the end of the current browser event loop.

**Comment by kyeotic** [comment]
I'm adding this to the v3 milestone, which I hope to have done in the next two weeks.

**Comment by roypearce** [comment]
> I'm adding this to the v3 milestone, which I hope to have done in the next two weeks.

😮 Nice! Thank you!

**Comment by roypearce** [comment]
Is there anything I can do to help with this?

**Comment by kyeotic** [comment]
Hey Roy, sorry for the delay on this, it fell off my radar. I got vaccinated and I've been spending a lot more time away from my computer lately 🥳

If you want to tackle this feature I wouldn't turn down a PR, and I'm happy to provide guidance on it. Otherwise I will pick it up next week sometime. 

I don't plan to hold this for v3 anymore, it's already taken too long.

**Comment by roypearce** [comment]
Congrats on the vaccine! I too have been making up for lost time thanks to it. 

I will try to put something together this weekend.

**Comment by roypearce** [comment]
I'm basically using the pseudo code that I wrote to test this out and have discovered an issue, and would need guidance on another thing. 

The issue is that `useLocationChange` is creating a new event listener for `popstate` every time it is called. It is being called multiple times from the render function of my app. Whether or not I use `useCallback` for the callback function, several event listeners are being added, and I'm not passing anything for the second argument.

This means if `shouldCancelNavigation()` is called from inside `onPopState` like in my pseudo code, then it is called for every event listener added, which means the confirm dialog shows up that many times in succession before moving on. 

I don't know if it is intentional to have multiple `popstate` listeners like in situations where `basePath` or `isActive` changes, it seems that way from the dependency arrays. So if there could be multiples, then we would need to only call it once per instance of popstate despite however many event listeners are fired from it, which adds some complication.

What I need guidance on is that the previous url needs to be known inside the `onPopState` function so it can use `pushState` to restore the url in the window if the user cancels navigation. I'm not sure if the last used url is something that is tracked right now, or how best you would want that tracked, but it seems this is not something obtainable from the event from the research I have done.

**Comment by kyeotic** [comment]
I'm a bit confused on how that psuedo code fits in here. You wrote it like its a hook, is it inside of `useNavigationPrompt`? Or is it replacing the similar code in `useLocationChange`?

If the latter, I don't think that's going to work. `useLocationChange` isn't necessarily in use when `useNavigationPrompt` is used, so "route-undo" logic there wouldn't necessarily be active. As you found out, it also might in use in multiple places, requires some rather awkward coordination. Even if it could be forced into working, I think `useLocationChange` is the wrong place to put that code.

I going to take a swing at implementing this inside `useNavigationPrompt` by patching the

**Comment by roypearce** [comment]
Your comment cut off so I missed the last bit...

But, yea, so I put it in the `onPopState` callback inside of `useLocationChange` because that is where the `popstate` event is firing, and it needs to be triggered from a `popstate` event because that is how we would know that the browser has navigated back from the browser back button. So it sounds like you would want to add another `popstate` handler inside of `useNavigationPrompt` and handle the logic there? That could be a lot simpler.

**Comment by kyeotic** [comment]
Yea sorry about that, not sure what happened.

Yes, I think it makes more sense to wire this up in `useNavigationPrompt`. I tried overwriting the `history.pushState` function, but its not invoked when the back button is hit in Firefox. This probably has something to to with their [back-forward cache](https://web.dev/bfcache/). I don't see a good way to tell that the back button was invoked when the previous route is activated.

**Comment by roypearce** [comment]
I just had a play with adding it to `useNavigationPrompt` 
```
export function useNavigationPrompt(predicate = true, prompt = defaultPrompt) {
  if (isNode) return
  
  useLayoutEffect(() => {
    const onPopStateNavigationPrompt = () => {
      if (shouldCancelNavigation()) {
        window.history.pushState(null, null, lastPath)
        return
      }
    }
    window.addEventListener('popstate', onPopStateNavigationPrompt)

    const handler = e => {
      if (predicate) {
        return e ? cancelNavigation(e, prompt) : prompt
      }
    }
    addInterceptor(handler)

    return () => {
      window.removeEventListener('popstate', onPopState2)
      removeInterceptor(handler)
    }
  }, [predicate, prompt])
}
```

In theory that could work, in practice the event listener here gets removed before it has a chance to fire. What happens is the `popstate` listener in `useLocationChange` fires first, the changes from that cascade and unmount `useNavigationPrompt`, removing the `interceptors` and its event listener.

**Comment by kyeotic** [comment]
Maybe it needs to be put in a 0ms timeout?

**Comment by roypearce** [comment]
I added a timeout and while that works for my use case it causes all the tests to fail. Maybe you have some insights, or thoughts on an alternative solution. https://github.com/roypearce/raviger/pull/new/raviger-issue-81

**Comment by kyeotic** [comment]
The tests fail because they expect navigation to be synchronous, so they aren't waiting for the timeout before making their assertions.

It does confirm your theory though. Maybe instead of that we could put the cleanup in a timeout, so that navigation still fired synchronously.

**Comment by roypearce** [comment]
I think the `setTimeout` need to be [here](https://github.com/kyeotic/raviger/compare/main...roypearce:raviger-issue-81?expand=1#diff-4e7e970f4382ec4d3d40f70e49bbce5249d95f7ecfe4603ac47741f89d94716eR83) because [this](https://github.com/kyeotic/raviger/blob/main/src/path.js#L86) triggers the cascade of events that needs to be blocked while the events from `useNavigationPrompt` have a chance to fire. That's because there are two specific things happening for router navigation, as you know, the url is getting updated, and React is rendering a different component. The url we know we can't stop from changing due to not being able to intercept it, but we do need to stop the new page from rendering while the prompt is being resolved. That's because if the new page renders then the current state of the page the user may want to stay on is lost, and we're trying to prevent that unless they want to continue.

The `popstate` events for `useNavigationPrompt` being unregistered before they had a chance to fire was just a symptom of the issue, moving the `setTimeout` elsewhere just results in undesirable outcomes. 

I did try to see if `interceptors.size` was truthy then run the `onPopState` callback in a `setTimeout` while letting the other instances run asynchronously, but that caused even more tests to fail. 

If the branch I shared is the best option for fixing the issue, do you want to help update the tests to work async? It looks like `await wait(expect(...))` syntax works for the current version of `testing-library` this project is on. 

If you want to go another direction, let me know how I can help.

**Comment by kyeotic** [comment]
So I want to avoid a timeout if at all possible. Keeping navigation synchronous is very valuable since React renders synchronously.

I think interception needs some extra work to do this properly. As you noted, if `shouldCancelNavigation` is in `useLocationChange` then multiple prompts appear. If there isn't a check in `useLocationChange` then the content will get unloaded, since thats what controls the url update that cascades back up to `useRoutes`.

Having it in both places blocks navigation from the back button with a prompt, but the prompt appears multiple times. This works even without any timeouts. The check needs to happen and it needs to be deduped. Intercept may need to track its prompts somehow; the issue is when to reset them.

I'm tinkering with this now.

**Comment by roypearce** [comment]
Have you had any luck?

**Comment by kyeotic** [comment]
published under 2.3.0

**Comment by roypearce** [comment]
Awesome, thank you so much!
